### ISSUE TITLE
The config.yaml settings in experiments/ are quite different from the settings in the paper

### ISSUE THREAD

**Issue by jianjieluo** [locuslab/mdeq#1]
@jerrybai1995 @vkoltun @zkolter 

Thanks for the nice work! I am quite interested in the MDEQ model and try to train it on the cifar10 dataset. However, I can't get the expected performances reported in the paper. Specifically, the accuracy of `MDEQ-small(ours)` without data augmentation in Table 1 is 87.1% while I only get 80.3% after removing the data augmentation code in `tools/cls_train.py`.

I checked the `experiments\cifar\cls_mdeq_TINY.yaml` and `experiments\cifar\cls_mdeq_LARGE.yaml` carefully and found that the settings are quite different from the settings in the paper (i.e. Table 4 in Appendix A), including dropout rate, For-Backword Thresholds, group num of GroupNormalization and so on. I have adjusted the setting in Table 4 while the performance is not improved. I have no idea with whether the  `LR_STEP` or other settings in the `.yaml` file which can not be found in the paper harm the training process. 

Is there any suggestion to reimplement the performance(87.1% ± 0.4%) in Table 1 in the paper? In addition, I will appreciate that if you can share the YAML files used in your paper experiments since I find the `config.yaml` settings of `cifar10` and `imagenet` in `experiments/` are quite different from the settings in Table 4 in Appendix A of the original paper.

**Comment by shaojieb** [comment]
Hi @jianjieluo ,

Thanks for the question! I just tried to reproduce it and am able to get >85% by adopting a larger dropout rate, adding a very small weight decay, and reducing the # of pretraining layers (but I lost the original script due to code clean up, so I did try a few times indeed, see my discussion on overfitting below). 

Remember that CIFAR-10 **without** data augmentation is extremely small and very easy to overfit on. Such sensitivity to the regularization hyperparameters is a strong piece of evidence, which is why very few works use it. One thing you can also try, and should be very useful, is to inject the input to all resolutions (this does not regularize the model, but essentially better train each equilibrium stream, and adds very minimal # of parameters; you only need to add a convolution for each downsampling. You can do this by modifying the `self.stage0` in `mdeq_core.py` on L377... let me know (shaojieb@cs.cmu.edu) if you want me to provide a code sample that does this).  But more broadly, I highly encourage you to run the CIFAR-10 **with** data augmentation setting (it's only about 4x more epochs) but the results are generally stabler, more indicative, and less subject to the overfitting problem (It will overfit in the end still). ImageNet is better, of course, but takes longer to run.

Also, in general, for CIFAR experiments you don't need the "30" F/B_THRESHOLD... something smaller would work just fine (and make your program run a lot faster).

And yes, It is indeed true that the yaml files were a bit different--- I believe the ones provided in the repo are in fact generally better than the ones in the paper (and we'll update some of the numbers in the results table & Table 4 in the next version of the paper).  But thanks for your suggestion, I'll check each of these scripts and update the config files to make sure they are more consistent in the next release.

**Comment by jianjieluo** [comment]
Hi @jerrybai1995 ,
Thanks for your quick response and detailed suggestions! 💯 

Would you please share your whole >85% settings here **with accurate number**(i.e. `cls_mdeq_TINY.yaml`)? I am trying the CIFAR-10 without data augmentation experiment with `DROPOUT=0.15->0.3`, `WD=0->5e-5`(meanwhile adding `weight_decay` param for adam solver in `get_optimizer()` function), `NUM_LAYERS=10->5` and `F/B_THRES=30->18` followed your suggestions, but the valid loss/top1 curves seem to be even a bit worse(<80%) than before, though the training process is still running (20/50 epoch). It's hard for me to tune the hyper-parameters to mitigate the overfitting problem in a short time with quite limited computing resources :-(.

In addition, I will try to run MDEQ on CIFAR-10 with data augmentation setting and on ImageNet once I fix the CIFAR-10 without data augmentation setting. Thanks for your patient explanation again.

Best Wishes.

**Comment by shaojieb** [comment]
Hi @jianjieluo ,

Sorry for the (a bit) delayed response. I was using my own version of the code (which contains some new stuff and is not as clean :P) and after trying out this public version of the repo, I did indeed encounter a similar problem that you reported above.

Delving into this issue, I found it is not exactly a hyperparameter one--- but instead my oversight when cleaning up the code for public release ~2 months ago (when I focused more on ImageNet and Cityscapes and their reproducibility). Specifically, I simplified the input injection transformation (an explicit part of the model, before MDEQ) in CIFAR-10 (no data augmentation) because it was prone to overfitting. However, I forgot to include this simplification in this public version.

I've updated the repo accordingly. At the same time, I found that 2 scales are already sufficient for reproducing >85% accuracy on CIFAR-10 (without augmentation), and have pushed a new `cls_mdeq_TINY.yaml` along the way in this commit (I also added an "AUGMENT" entry to the config setting so that you can use it to directly control data augmentation 😃). Please pull the repo again and run
```sh
python tools/cls_train.py --cfg experiments/cifar/cls_mdeq_TINY.yaml
```
But still, as I promised above, we'll also update the hyperparameter table for CIFAR-10 (no augmentation) in the next version of the paper as well. Thanks for pointing out the problem! Please let me know if this helps.

(P.S.: I would still encourage you to try the CIFAR-10 **with augmentation** setting... as you can see, such simplification of input injection that I missed is important in the no augmentation setting of CIFAR-10, but would no longer matter in the with augmentation setting. You should be able to get >93.5% accuracy in CIFAR-10 pretty consistently.)

**Comment by jianjieluo** [comment]
Hi @jerrybai1995 ,

Thanks for the bug fix and I will try it right now!

I have tried the CIFAR-10 **with augmentation setting** recently, and I get ~93.34% accuracy which is reasonable. Thanks for your advice.

Besides, I find that maybe you need to add the `weight_decay` param for `Adam` solver in the `get_optimizer()` function in `lib/utils/utils.py`, otherwise your `WD` setting in CIFAR-10 `config.yml` files won't be used in the actual training process, although it may be not related to this issue.

I will close this issue once I get a reasonable performance on the CIFAR-10 `cls_mdey_TINY` setting. Thank you for the quick and detailed response again and **it does help me a lot**.

Best Wishes.

**Comment by shaojieb** [comment]
Ah, good point about the WD! Another thing that I overlooked when I cleaned up the code. Thanks a lot for the pointer.

And yes--- do let me know if you still can't improve over 80% on CIFAR-10 (no augmentation) 😄

**Comment by jianjieluo** [comment]
After adding the `weight_decay` param discussed above and trying with the newest code & config(22731498bcd5524976664000367caa179fe6c56e), I can finally get ~84.5% accuracy on CIFAR-10 **without augmentation**, which is reasonable according to jerrybai1995's response. I will close this issue.

Thanks for the quick response and bug fix again 💯.
